src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs

```diff
@@ -68,6 +70,31 @@ public SpanMapper(ITextBuffer primaryBuffer)
                 /// </summary>
                 public bool SupportsMappingImportDirectives => false;
 
+                public async Task<ImmutableArray<(string mappedFilePath, TextChange mappedTextChange)>> GetTextChangesAsync(
+                    Document oldDocument,
+                    Document newDocument,
+                    CancellationToken cancellationToken)
+                {
+                    var textChanges = (await newDocument.GetTextChangesAsync(oldDocument, cancellationToken).ConfigureAwait(false)).ToImmutableArray();
+                    var mappedSpanResults = await MapSpansAsync(oldDocument, textChanges.Select(tc => tc.Span), CancellationToken.None).ConfigureAwait(false);
+
+                    Contract.ThrowIfFalse(mappedSpanResults.Length == textChanges.Length);
+
+                    using var _ = ArrayBuilder<(string, TextChange)>.GetInstance(out var mappedFilePathToTextChange);
+                    for (var i = 0; i < mappedSpanResults.Length; i++)
+                    {
+                        // Only include changes that could be mapped.
+                        var newText = textChanges[i].NewText;
+                        if (!mappedSpanResults[i].IsDefault && newText != null)
+                        {
+                            var newTextChange = new TextChange(mappedSpanResults[i].Span, newText);
+                            mappedFilePathToTextChange.Add((mappedSpanResults[i].FilePath, newTextChange));
+                        }
+                    }
+
+                    return mappedFilePathToTextChange.ToImmutable();
+                }
```

Comment:
it would be nice if we could take the common portion of this code (and what's in the razor versin) and put that into a shared location.

Reply:
I abstracted out a class and put the common portion there, hopefully that improves things!